### DISCUSSION TITLE
Create albums for top 10 unnamed persons

### DISCUSSION THREAD

**Discussion by jefgoi** [RhetTbull/osxphotos#1487] [Q&A] [answered]
Hello. 
The goal: to name most of persons in the library.

The context.
I have about 100k photos in the library. It’s more than 100 persons, each 20-50-100+ photos. Photos app shows only some of persons in Persons&Pets section - just about 10 (even after weeks of analyzing all photos). I can discover more persons manually by “travelling” from person to person, from photo to photo. With this approach Photos app shows me all photos where this specific unnamed person was identified. Then I can name it.

The question.
How can I create with osxphotos 10 albums for top 10 unnamed persons (by number of photos)? It will help a lot: I will just name the persons in each album.

**Comment by RhetTbull** [comment]
Hi @jefgoi It's not currently possible to do what you're asking but I believe it would be possible with a custom query function. I might be able to take a look at this later in the week.  It shouldn't be difficult as each face has a unique ID and so the query would need to find all faces with no names then group these. 

In the meantime, the following will find all photos with unnamed faces and put them in the album "Unnamed Faces":

`osxphotos query --query-function https://raw.githubusercontent.com/RhetTbull/osxphotos/main/examples/find_unnamed_faces.py::unnamed_faces --add-to-album "Unnamed Faces"`

This downloads and runs the custom query function found [here](https://github.com/RhetTbull/osxphotos/blob/main/examples/find_unnamed_faces.py).

**Reply by jefgoi** [reply]
Thanks! I’m already exploring this fantastic possibility to use own python code - this is awesome! (E.g., I use it to find all photos with just two faces with good quality, one of which is already named and another is not. It helped a lot to find some diamonds in my family photo library.)

As for the grouping by ID, it’s exactly what I’m thinking about, but I don’t know what are those tables and fields with ID I need… 

Thanks in advance, looking forward to hearing from you about such possibility!

PS What I managed to find so far:
1) Sort ZPERSON by ZFACECOUNT - it gives most popular persons, even unnamed!

2) Select all ZDETECTEDFACE.ZPERSONFORFACE by ZPERSON.Z_PK from (1) - gives the list of faces assigned to specific person

3) I could not find the table where many to many ZDETECTEDFACE.Z_PK to ZASSET.Z_PK are stored to find specific image for each face.

**Reply by RhetTbull** [reply]
The PhotoDB.person_info property will give you a list of [PersonInfo](https://rhettbull.github.io/osxphotos/API_README.html#id144) objects representing the people in the library. Each person will have a unique ID (uuid) so you can group persons by uuid. You can search for those that don't have a name defined then use the `PersonInfo.photos` property to get a list of all photos that person is in.

**Reply by RhetTbull** [reply]
Something like this should work to get the top 10:

```python
"""Find unnamed faces by photo count"""

import osxphotos

def main():
    photosdb = osxphotos.PhotosDB()
    unnamed_persons = {}
    for person in photosdb.person_info:
        if person.name == "_UNKNOWN_":
            try:
                unnamed_persons[person.uuid] = len(person.photos)
            except KeyError:
                pass
    top10 = sorted(unnamed_persons.items(), key=lambda x: x[1], reverse=True)[:10]
    print(top10)

if __name__ == "__main__":
    main()
```

I had to put the try/except in there because there are people that don't have photos (for example, the photos were deleted but the person entry is still in the database). These caused an error when calling `PersonInfo.photos`

Also, "\_UNKNOWN\_" is a marker inserted by OSXPhotos for persons with no name. It's defined in `osxphotos._constants` as `_UNKNOWN_PERSON` if you want to import it programmatically. Using a known marker simplifies some of the code elsewhere in OSXPhotos.

**Reply by RhetTbull** [reply]
In general, with osxphotos you should not have to resort to the underlying SQL database when using OSXPhotos as the goal is to expose this data for you in a pythonic interface. Of course, there are some things that aren't implemented but you should find that almost everything you need is already exposed in OSXPhotos. See the [API documentation](https://rhettbull.github.io/osxphotos/API_README.html#)

**Reply by RhetTbull** [reply, marked as answer]
I decided this might actually be useful for me too so I whipped up a script that will find the top 10 unnamed faces and add them to albums automatically:

```python
"""Find top 10 unnamed faces in Photos library and add them to albums."""

from osxphotos import PhotosDB
from osxphotos._constants import _UNKNOWN_PERSON
from osxphotos.photosalbum import PhotosAlbum


def get_person_photos(person_info):
    """Workaround for bug #1489 in osxphotos, will be fixed in next release"""
    try:
        return person_info.photos
    except KeyError:
        return []


def main():
    """Find top 10 unnamed faces in Photos library and add them to albums."""
    photosdb = PhotosDB(verbose=print)
    unnamed_persons = [p for p in photosdb.person_info if p.name == _UNKNOWN_PERSON]
    print(f"Found {len(unnamed_persons)} unnamed persons")

    top_10_unnamed = sorted(
        unnamed_persons, key=lambda x: len(get_person_photos(x)), reverse=True
    )[:10]

    for p in top_10_unnamed:
        photos = get_person_photos(p)
        album_name = f"Unnamed person: {p.uuid} ({len(photos)})"
        print(f"Creating album '{album_name}'")
        album = PhotosAlbum(album_name)
        album.add_list(photos)

if __name__ == "__main__":
    main()

```

This can be run directly from osxphotos via `osxphotos run top_10_faces.py` or via python if you've installed the osxphotos package.

**Reply by jefgoi** [reply]
Fantastic!!! Thanks a lot!

PS Made small adjustment: I add only first 50 photos to the album. It’s enough to tag the person, and it works X times faster.

**Reply by Conzpiral** [reply]
Thank you, this is great.
I don't know if this is the right place to ask but...
My Library has a lot of those old iPhoto manually added faces without names. I think is the same as the one if you manually add face and the adjustable circle says somthing like 'click to give name' (my Photo app is in norwegian so guessing).
My question is if there is a way to find all those old manually randomly added circles with no name which stay behind from years ago when the app was called iPhoto?

**Reply by RhetTbull** [reply]
Hi @Conzpiral The following should work. This turned out to be a bit more complex that it should be because I discovered osxphotos doesn't give you access to manually added faces with no name (opened new issue #1493 to fix this). The following script which can be saved to "manually_added_faces.py" or downloaded from [here](https://raw.githubusercontent.com/RhetTbull/osxphotos/main/examples/manually_added_faces.py) can be run with `osxphotos run manually_added_faces.py` to find all manually added faces to the album "Manual Faces". To find only unnamed manual faces, add the `-u` flag: `osxphotos run manually_added_faces.py -u` and to add to a specific album, use the `-a` flag: `osxphotos run manually_added_faces.py -u -a "Manual Faces With No Name'` . The album does not need to already exist -- it will be created if needed.

You can also have osxphotos download the code and run it for you without saving it locally:

`osxphotos run https://raw.githubusercontent.com/RhetTbull/osxphotos/main/examples/manually_added_faces.py -u -a 'Album Name'`

```python
"""Find manually added faces in Photos library and add them to an album."""

import click

from osxphotos import PhotosDB
from osxphotos._constants import _DB_TABLE_NAMES
from osxphotos.photosalbum import PhotosAlbum


def get_person_photos(person_info):
    """Workaround for bug #1489 in osxphotos, will be fixed in next release"""
    try:
        return person_info.photos
    except KeyError:
        return []


@click.command()
@click.option(
    "-u", "--unnamed", is_flag=True, help="Find only unnamed manually added faces"
)
@click.option(
    "-a", "--album", help="Add photos to this album; default = 'Manual Faces'"
)
def main(unnamed: bool, album: str):
    """Find photos with manually added faces and add them to an album in Photos."""
    photosdb = PhotosDB(verbose=print)
    photos = photosdb.photos()

    if unnamed:
        # Currently, osxphotos doesn't provide API access to manually added faces with no associated person
        # So need to query the database directly
        photos_version = photosdb.photos_version
        asset_table = _DB_TABLE_NAMES[photos_version]["ASSET"]
        asset_fk = _DB_TABLE_NAMES[photos_version]["DETECTED_FACE_ASSET_FK"]
        results = photosdb.execute(
            f""" SELECT {asset_table}.ZUUID
                FROM {asset_table}
                INNER JOIN ZDETECTEDFACE
                ON {asset_table}.Z_PK = {asset_fk}
                WHERE ZDETECTEDFACE.ZMANUAL = 1 AND ZDETECTEDFACE.ZPERSON IS NULL;
            """
        )
        uuids = [r[0] for r in results]
        manual_faces = photosdb.photos(uuid=uuids)
        print(f"Found {len(manual_faces)} photos with unnamed manually added faces")
    else:
        manual_faces = [p for p in photos for f in p.face_info if f.manual]
        print(f"Found {len(manual_faces)} photos with manually added faces")

    if album is None:
        album = "Manual Faces"

    print(f"Adding photos to album '{album}'")
    album = PhotosAlbum(album)
    album.add_list(manual_faces)


if __name__ == "__main__":
    main()
```

**Reply by Conzpiral** [reply]
Thank you so much, I will try it now, this is one thing bothering me for years. Osxphotosis is incredible and have helped me out so much fixing my old Libraies, can't thank you enough, you are the best!

**Reply by Conzpiral** [reply]
Ok, one more question, hope it is ok to post it here.
At the bottom of the Persons in Photos app it seemed to have stopped and for 3 years now it still says it has 227 pictures remains to search. Is there a way to find which 227 pictures/videos this is so I can investigate those for errors?

**Reply by RhetTbull** [reply]
@Conzpiral I'm not 100% certain this will do the trick but I think it's a good start.  When the photo analysis process runs on a photo it updates a date stamp of when the analysis took place. OSXPhotos doesn't provide a way to directly access this but it can be queried using the `tables()` method which lets you access the underlying database tables. The following will run very slowly as it's got to run a query for every single photo in the library but it will find all photos that have not been analyzed. You can add `--count` just to see the total count or `--add-to-album "Album name"` to add the found photos to a given album.

`osxphotos query --not-shared --query-eval "not photo.shared_moment" --query-eval "photo.tables().ZASSET.ZANALYSISSTATEMODIFICATIONDATE[0] is None"`

The `--not-shared` and `--query-eval "not photo.shared_moment"` exclude images from shared albums and shared moments (photos shared as a group via Messages) which won't be analyzed. 

If this doesn't give a result close to what you're expecting let me know and I'll dig further.

**Reply by RhetTbull** [reply]
@Conzpiral I've created a new counting utility to count all sorts of things in Photos including the number of unanalyzed faces. See the thread at #1499 for instructions on how to run it.

**Reply by Conzpiral** [reply]
I ran the osxphotos query --not-shared --query-eval "not photo.shared_moment" --query-eval "photo.tables().ZASSET.ZANALYSISSTATEMODIFICATIONDATE[0] is None" --add-to-album "Album name". There was no photos.

But then I ran your new counting utility script and found 'manually added faces: 227'. This is exactly the number I was looking for. I guess this means Photos app don't recognize those as real faces, and I think I know why. This is because sometimes I just manually added the face even when it e.g. could be my grandfather without face from behind, just to sort of 'tag' the photo.

Thank you so much, my messy Library know starts to get very clean thanks to OSXPhotos. You have saved so many days for me!

When I run osxphotos run https://raw.githubusercontent.com/RhetTbull/osxphotos/main/examples/manually_added_faces.py -a 'Manual Faces' without the -u flag terminal says it found 119 photos but the album created shows 111 photos in Photos.

I will look further into this :-)

**Reply by Conzpiral** [reply]
Ok, so I ran osxphotos run https://raw.githubusercontent.com/RhetTbull/osxphotos/main/examples/manually_added_faces.py -a 'Manual Faces'  to create an album. I exported those originals out of Photos and deleted them in the app. Now when I run the counts.py script it says there is 108 manually added faces left. Photos app now at the same time shows 216 the exact double of this left counting in the faces inside Photos app. I don't know where to go from here, how do I find those 'manually added faces: 108' found by the counts.py script?

**Reply by RhetTbull** [reply]
@Conzpiral sorry, I don't have time today to dig into this. The face count shown in the `counts.py` script is the count of *faces* not photos with those faces.  The faces are separate entities in the database and deleting the photos does not delete them from the database. They live in the database forever. The `manually_added_faces.py` script on the other hand, counts *photos* with manually added faces which is a different count. They're counting different things.
